File: lib/services/google_calendar.rb
# file lib/services/google_calendar.rb

require 'active_support/time'
require 'httparty'

module Services
  class GoogleCalendar
    include HTTParty

    base_uri 'https://www.googleapis.com/calendar/v3/calendars'

    def initialize(api_key)
      @api_key = api_key
    end

    def get_calendar(calendar_id, options = {})
      get_calendar_entries(calendar_id, options)
    end

    private

    attr_accessor :api_key

    def get_calendar_entries(calendar_id, options = {})
      defaults = {
        query: {
          key: @api_key,
          singleEvents: true,
          orderBy: 'startTime',
          timeMin: DateTime.now,
          timeMax: DateTime.now + 48.hours
        }
      }
      options = defaults.merge(options)
      response = self.class.get("/#{calendar_id}/events", options)
      JSON.parse(response.body)['items']
    end
  end
end

File: src/services/google-calendar.js
import request from 'request'

/**
 * GoogleCalendar class
 *
 * @export
 * @class GoogleCalendar
 */
export default class GoogleCalendar {
  /**
   * Creates an instance of GoogleCalendar
   *
   * @param {string} calendarId
   * @param {string} apiKey
   *
   * @memberOf GoogleCalendar
   */
  constructor (calendarId, apiKey) {
    this.calendarId = calendarId
    this.apiKey = apiKey
  }

  /**
   * Retrieves events from calendar
   *
   * @param {Date} startTime
   * @param {Date} endTime
   * @returns {PromiseLike<Array<Object>>}
   *
   * @memberOf GoogleCalendar
   */
  getEvents (startTime, endTime) {
    let url = 'https://www.googleapis.com/calendar/v3/calendars/'
    url += `${this.calendarId}/events/`
    url += `?key=${this.apiKey}`
    url += `&timeMin=${startTime.toISOString()}`
    url += `&timeMax=${endTime.toISOString()}`
    url += '&singleEvents=true'
    url += '&orderBy=startTime'

    return new Promise((resolve, reject) => {
      request(url, (error, response, body) => {
        if (error) {
          reject(error)
          return
        }
        resolve(JSON.parse(body).items)
      })
    })
  }
}
